### PR TITLE
Issue 81: Customize pravega payload attributes before submitting to make each payload unique

### LINKED ISSUE #81: Custom payload runtime changes for pravega
**Problem description**
We need to add some changes to customize request payload to make it unique by modifying it's attributes.

Customizing request payload sample to create unique request payload for each submitted message of pravega.

### PR DESCRIPTION
**Pull request by dada-dell-emc**
**Change log description**
Added changes to customize pravega payload **Timestamp** attribute before submitting to Pravega, to make each payload unique.

**Purpose of the change**
Fixes https://github.com/pravega/openmessaging-benchmark/issues/81.

**What the code does**
This PR does the following:
Introduced new boolean configuration attribute **_customPayload_**.
If **customPayload** is set to true, then customize json request payload with Timestamp dynamic value in ms.

**How to verify it**
Enable  **customPayload** configuration for pravega and then run omb for pravega driver with sample payload **custom-payload-idrac.data** and workload **custom-payload-idrac.yaml** shared in this repository.
After that when you read those messages in pravega consumer check **Timestamp** attribute value, you will get different values almost each time in milliseconds. Or you can enble debug log in omb which will print customized payload for pravega, check **Timestamp** attribute value,it will print different values almost each time in milliseconds.

### CODE REVIEW

File: driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+package io.openmessaging.benchmark.driver.pravega;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Map;
+
+public class CustomPayloadUtils {
```
Comment (shshashwat) [inline driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java:28]: Public classes should have javadoc

File: driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    private static final Logger log = LoggerFactory.getLogger(CustomPayloadUtils.class);
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    public static byte[] customizePayload(boolean isCustomPayload, byte[] payload) {
+        if(isCustomPayload && payload!=null && payload.length > 0) {
```
Comment (shshashwat) [inline driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java:33]: can something else be tried like `ArrayUtils.isNotEmpty(payLoad)`

File: driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    /**
+     * @param isCustomPayload custom payload is enabled or not
+     * @param payload static default payload to send to messaging system.
+     * @return customized payload in byte array
+     */
+    public static byte[] customizePayload(boolean isCustomPayload, byte[] payload) {
+        if(isCustomPayload && ArrayUtils.isNotEmpty(payload)) {
+            try {
+                Map<String, Object> jsonPayload = mapper.readValue(payload, Map.class);
+                jsonPayload.put("Timestamp",System.currentTimeMillis());
```
Comment (ShwethaSNayak) [inline driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java:45]: Is there a possibility to have timestamp object in incoming JSON payload?

**Comment by ShwethaSNayak** [comment]
Did we verify these changes through any tests?

File: driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/PravegaBenchmarkTransactionProducer.java [hunk 4]
```diff
@@ -59,10 +59,16 @@ public PravegaBenchmarkTransactionProducer(String streamName, EventStreamClientF
         this.eventsPerTransaction = eventsPerTransaction;
         this.includeTimestampInEvent = includeTimestampInEvent;
     }
+    public PravegaBenchmarkTransactionProducer(String streamName, EventStreamClientFactory clientFactory,
```
Comment (sachin-j-joshi) [inline driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/PravegaBenchmarkTransactionProducer.java:62]: please add comment to public methods. Also add empty line between definitions.

File: driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java (hunk 3 above)
Comment (dada-dell-emc) [reply driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java:45]: Yes, main intension here is to make each payload unique, and Timestamp attribute is easier than other object to make payload unique.

**Comment by dada-dell-emc** [comment]
> Did we verify these changes through any tests?

We verified these changes from logs, there is no test cases for drivers

File: driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    private static final Logger log = LoggerFactory.getLogger(CustomPayloadUtils.class);
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    public static byte[] customizePayload(boolean isCustomPayload, byte[] payload) {
+        if(isCustomPayload && payload!=null && payload.length > 0) {
+            try {
+                Map<String, Object> jsonPayload = mapper.readValue(payload, Map.class);
+                jsonPayload.put("Timestamp",System.currentTimeMillis());
+                payload = mapper.writeValueAsBytes(jsonPayload);
+                if(log.isDebugEnabled()) {
```
[the remark this answers is not in the document]
Comment (shshashwat) [reply driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java:47]: @dada-dell-emc, please provide your answer before marking it as resolved

**Review by sachin-j-joshi — APPROVED** [review]
LGTM

**Review by shshashwat — APPROVED** [review]

File: payload/custom-payload-idrac.data [hunk 6]
```diff
@@ -0,0 +1 @@
...3.61.103","host":"telegraf-polling-service","system-name":"iDRAC-3XJJS13"},"timestamp":1686258320},{"fields":{"cmos-batterystate":4,"cpu1-temp":180,"cpu2-temp":380,"exhaust-temp":680,"fan1-speed":6480,"fan2-speed":5160,"fan3-speed":6600,"fan4-speed":5160,"fan5-speed":6600,"fan6-speed":5160,"idrac-url":"https://10.243.61.107:443","inlet-temp":580,"power-state":4,"system-globalstatus":3,"system-model":"PowerEdge R640","system-servicetag":"3XKPH13","system-uptime":17300885,"system-watts":312},"name":"snmp","tags":{"agent_host":"10.243.61.107","host":"telegraf-polling-service","system-name":"iDRAC-3XKPH13"},"timestamp":1686258320},{"fields":{"cmos-batterystate":4,"cpu1-temp":190,"cpu2-temp":330,"exhaust-temp":490,"fan1-speed":11280,"fan2-speed":8760,"fan3-speed":11280,"fan4-speed":8640,"fan5-speed":11160,"fan6-speed":8640,"idrac-url":"https://10.243.61.104:443","inlet-temp":440,"power-state":4,"system-globalstatus":3,"system-model":"PowerEdge R640","system-servicetag":"3XHPH13","system-uptime":2499984,"system-watts":312},"name":"snmp","tags":{"agent_host":"10.243.61.104","host":"telegraf-polling-service","system-name":"iDRAC-3XHPH13"},"timestamp":1686258320}],"Timestamp":1686258320156}
```
Comment (RaulGracia) [inline payload/custom-payload-idrac.data:1]: This seems to be leaking IPs in a public repo, is this ok?

File: payload/custom-payload-k8s.data [hunk 7]
```diff
@@ -0,0 +1 @@
..."app_kubernetes_io_component":"metrics","app_kubernetes_io_instance":"cluster-monitoring","app_kubernetes_io_managed_by":"Helm","app_kubernetes_io_name":"kube-state-metrics","app_kubernetes_io_part_of":"kube-state-metrics","app_kubernetes_io_version":"2.7.0","configmap":"project-operator-app-config","helm_sh_chart":"kube-state-metrics-4.17.0","host":"telegraf-polling-service","instance":"10.233.102.154:8080","job":"kubernetes-service-endpoints","kubernetes_name":"cluster-monitoring-kube-state-metrics","kubernetes_namespace":"cluster-monitoring","kubernetes_node":"node1","namespace":"nautilus-system"},"timestamp":1686270341},{"fields":{"kube_configmap_info":1},"name":"prometheus_remote_write","tags":{"app_kubernetes_io_component":"metrics","app_kubernetes_io_instance":"cluster-monitoring","app_kubernetes_io_managed_by":"Helm","app_kubernetes_io_name":"kube-state-metrics","app_kubernetes_io_part_of":"kube-state-metrics","app_kubernetes_io_version":"2.7.0","configmap":"installer-logs-2023-06-03.03-48-49","helm_sh_chart":"kube-state-metrics-4.17.0","host":"telegraf-polling-service","instance":"10.233.102.154:8080","job":"kubernetes-service-endpoints","kubernetes_name":"cluster-monitori
```
Comment (RaulGracia) [inline payload/custom-payload-k8s.data:1]: Is this data safe to be published?

File: driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+public class CustomPayloadUtils {
+    private static final Logger log = LoggerFactory.getLogger(CustomPayloadUtils.class);
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    /**
+     * @param isCustomPayload custom payload is enabled or not
+     * @param payload static default payload to send to messaging system.
+     * @return customized payload in byte array
+     */
+    public static byte[] customizePayload(boolean isCustomPayload, byte[] payload) {
```
Comment (RaulGracia) [inline driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java:41]: The name of this method is not entirely describing what it does. It is actually generating a unique payload by means pf changing one field of the initial static payload. The question is: what is expected to be achieved from this?

**Review by RaulGracia — COMMENTED** [review]
If the intention of this PR is just to make payloads unique on a per-write basis, please consider a more efficient approach, as the current one looks quite expensive for a benchmark. Also, please note my comment about contributing in the future (maybe not this change, but future improvements such as upgrading dependencies) to the official OpenMessaging Benchmark repo, as we contributed the Pravega driver there a couple of years back: https://github.com/openmessaging/benchmark/tree/master/driver-pravega

File: driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java (hunk 8 above)
Comment (dada-dell-emc) [reply driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java:45]: Added doc comments to explain method functioning.

File: driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        if(isCustomPayload && ArrayUtils.isNotEmpty(payload)) {
+            try {
+                Map<String, Object> jsonPayload = mapper.readValue(payload, Map.class);
+                jsonPayload.put("Timestamp",System.currentTimeMillis());
+                payload = mapper.writeValueAsBytes(jsonPayload);
```
[the remark this answers is not in the document]
Comment (dada-dell-emc) [reply driver-pravega/src/main/java/io/openmessaging/benchmark/driver/pravega/CustomPayloadUtils.java:50]: We have room for improvement here, these changes are mainly to submit payloads into pravega for QE's performance measurement.
I will definitely work on making this efficient in my next code change.

**Comment by dada-dell-emc** [comment]
> If the intention of this PR is just to make payloads unique on a per-write basis, please consider a more efficient approach, as the current one looks quite expensive for a benchmark. Also, please note my comment about contributing in the future (maybe not this change, but future improvements such as upgrading dependencies) to the official OpenMessaging Benchmark repo, as we contributed the Pravega driver there a couple of years back: https://github.com/openmessaging/benchmark/tree/master/driver-pravega

These changes are mainly just to submit the payload into pravega and not benchmarking pravega, so that QE can maintain throughput for their end to end testing.
We will improve efficiency of code for benchmark in our next PR.

**Review by abhinb — APPROVED** [review]

**Comment by sachin-j-joshi** [comment]
@dada-dell-emc I have merged this PR based on 3 approvals. Please work towards addressing issues 
 in https://github.com/pravega/openmessaging-benchmark/issues/82

**Pull request merged**